### DISCUSSION TITLE
Our usage of the word "stable"

### DISCUSSION THREAD

**Discussion by nbolton** [deskflow/deskflow#7837] [Development]
I believe the way we use the word "stable" is confusing to packagers (see https://github.com/teknowledgist/Chocolatey-packages/issues/108 and #7811).

We say "stable" on our readme to refer to v1.17.0; because of the false dichotomy (an overwhelmingly common bias), they assume that continuous build is _not_ stable. Looking at the recent packager comments, this can be quite confusing for package maintainers.

I don't think we can guarantee that the continuous build is "stable" by most people's definition because it is `master` which fluxuates in stability. Continuous could possibly be stable, unstable or something else.

@nicola02nb @marlop352 @teknowledgist @cdwmhcc @SelfRef

**Comment by SelfRef** [comment]
I have an idea where misunderstanding may come from. On the releases page, the continuous build is incorrectly tagged as `Latest`. This tag is reserved for the latest stable version, which is 1.17.0. The solution for that would be to tag the continuous build as `Pre-release` to avoid confusion. It will show as:
![image](https://github.com/user-attachments/assets/8eb138ba-acbe-4f48-a217-b36475849665)

An additional advantage is that GitHub will show the most recent `Latest` tag on the main repo page, instead of the most recent `Pre-release`. In other words - this will show 1.17.0:
![image](https://github.com/user-attachments/assets/ac429d1d-573f-4384-ad8a-d480f0a5ddf3)

**Reply by cdwmhcc** [reply]
> An additional advantage is that GitHub will show the most recent `Latest` tag on the main repo page, instead of the most recent `Pre-release`. In other words - this will show 1.17.0:
> ![image](https://private-user-images.githubusercontent.com/69125050/381944457-ac429d1d-573f-4384-ad8a-d480f0a5ddf3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzNzkyNDEsIm5iZiI6MTczMDM3ODk0MSwicGF0aCI6Ii82OTEyNTA1MC8zODE5NDQ0NTctYWM0MjlkMWQtNTczZi00Mzg0LWFkOGEtZDQ4MGYwYTVkZGYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMxVDEyNDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhOTJhY2JkMmNhYjk3ZjY1NGM2ZTYxNjVlMjI4NjVmZDIyMWVkNjI4Y2EyNDc4YzY5NTIyZDlkZDUxNTE1ZDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4W1Sev8wC_mA0CGvv2ZdJu0xPrDTHMfyO-zFLlwhHr8)

Oh no! Now I have completely lost track of version numbers.
https://api.github.com/repos/deskflow/deskflow/releases/latest
tag_name => continuous
why?

**Reply by cdwmhcc** [reply]
> On the releases page, the continuous build is incorrectly tagged as `Latest`

I get it. But does it make sense that continuous builds don't provide version numbers?

**Reply by SelfRef** [reply]
The idea of continuous build is that it's not a proper release, but just a built of the current dev version. At least, I always understand it this way. As soon as the codebase is stable enough, it can be released as versioned release.

**Reply by cdwmhcc** [reply]
> but just a built of the current dev version.

No problem, no matter it is a proper release or a development version, my focus is on the version number. I want to provide the development version and the proper release of deskflow to users and let them choose for themselves, but the problem now is that I can't get the version number of the development version.

**Reply by SelfRef** [reply]
Same, that's why I suggested to not tag the continuous build as latest. I just look at the numbered version.

**Comment by sithlord48** [comment]
the continuous is our latest release  it is not a pre-release its a real release.

**Reply by SelfRef** [reply]
Is it always stable, tho? Because this is how [GitHub docs explain](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) release tagging:

> 10. Optionally, to notify users that the release is not ready for production and may be unstable, select This is a pre-release.

**Reply by sithlord48** [reply]
Yes

**Reply by nicola02nb** [reply]
So we can publish on winget the latest continuos release... Or am I wrong?

**Reply by SelfRef** [reply]
Now I'm lost. So what @nbolton says "Continuous could possibly be stable, unstable or something else." is true or not?

If it's built from the `master` branch, that sounds a little naive to always assume it's stable and package maintainers can treat it as a stable version - especially as it doesn't even have a proper version number.

**Reply by sithlord48** [reply]
1. Every build has a proper version number, Every Single (landed) commit has a unique version.

2. You can deploy the continuous but this is more difficult to verify as the sha is gonna change alot. Some places will  tumbleweed and sid may want to pull this tag for testing etc.. .

**Reply by fuzzzerd** [reply]
The continuous release maybe a "real" release, but because it has the potential to be

> stable, unstable, or something else

it only makes sense to use the GitHub pre-release flag on the release, because its the least common denominator. It is the primary signal to any consumer of the release whether or not its considered safe for their use. If the same commit that made a particular continuous release goes on to get a git tag and a full stable release later, that's OK and even expected.

It is your project and you can run it however you want; however, the way things are setup now coupled with the discussion in this thread, makes it clear you are going against the grain with respect to how most software projects operate when using GitHub releases. Again, that's your prerogative, but then you have the added burden of explaining it to everyone that expects projects on GitHub to conform to the those norms.

Projects that don't want to follow the GitHub release norms, publishing the releases directly on your own website, where you can better control the signals to users about which release is appropriate for them makes more sense than using GitHub releases as the publishing platform.

**Comment by marlop352** [comment]
@nbolton since you tagged me,
No sane packager would ever consider "continuous" stable, the tag itself is not stable and all binaries will vanish as soon as another commit to master is done.
I get you want people to use/test it so put that in the readme as the recommended release for users but "latest" on GitHub should only ever be used for tagged versions and the release artifacts should not be messed with unless there is a serious breaking bug or something like that than you may delete the artifacts and release a new tagged version. Don't ever rename or replace artifacts of tagged versions.

**Reply by marlop352** [reply]
Continuous also has the added problem that it only lists the changes from the last PR merge or commit on the release notes.

**Reply by cdwmhcc** [reply]
https://github.com/deskflow/deskflow/discussions/7837#discussioncomment-11111001

This should be a tag error:

1.17.0 => Latest
continuous => Pre-release

**Reply by marlop352** [reply]
@cdwmhcc I already tried to make them understand this before but they seem to not want to understand that, or they do and don't care 🤷🏻‍♂️

For whoever wants to read how it went last time: https://github.com/deskflow/deskflow/pull/7811#issuecomment-2437890193

**Reply by cdwmhcc** [reply]
> I already tried to make them understand this before but they seem to not want to understand that

Sorry, I noticed the discussion at https://github.com/deskflow/deskflow/discussions/7837#discussioncomment-11111024, which is confusing.

**Reply by marlop352** [reply]
> > I already tried to make them understand this before but they seem to not want to understand that
> 
> Sorry, I noticed the discussion at [#7837 (comment)](https://github.com/deskflow/deskflow/discussions/7837#discussioncomment-11111024), which is confusing.

No problem, maybe more people saying the same thing make them change their mind.

**Reply by sithlord48** [reply]
Do not use the `latest` tag from github. It just means the "newest" or most recently changed

**Reply by sithlord48** [reply]
you can Thumbdown that comment all you want but go look at how the latest tag is made

**Reply by cdwmhcc** [reply]
> Do not use the `latest` tag from github. It just means the "newest" or most recently changed

This was originally a very simple matter, but it has become so complicated.

- 1.7.0-beta.1
- 1.7.0-dev.1
- 1.7.0-rc.1
- 1.7.0-nightly.1
- 1.7.0

This is why I initially recommended using Semantic Versioning. It clearly describes the different stages of software versions.

However, the developers ultimately chose to use the versioning standard of `1.7.0.0`, and they provided some explanations. I think it's understandable not to use Semantic Versioning.

But seeing this discussion, I realized the issue is not so simple. At least there is a difference in understanding between how the developers perceive versions and how we do, leading to misunderstandings.

Common Software Versioning Categories:
- 'nightly'
- 'canary'
- 'dev'
- 'alpha'
- 'beta'
- 'rc'
- 'preview'
- 'stable'
- 'lts'
- 'production'

Normally, most people would consider 1.7.0 to be the stable version. If the term “continuous” needs to be included in the version naming, it would most closely align with the following types: nightly/canary/dev.

In the context of GitHub Releases, the term Latest is generally understood to mean the most recent stable release, not just the most recently updated or created. This is a key point of expectation for users navigating release pages.

Examples:
https://github.com/brave/brave-browser/releases
https://github.com/lapce/lapce/releases
https://github.com/desktop/desktop/releases
https://github.com/zed-industries/zed/releases

### Conclusion

Understanding the distinction between these different types of releases helps in setting the correct expectations for users and developers alike. It's important that the developers communicate clearly what each release signifies, especially if their versioning deviates from widely recognized standards like Semantic Versioning. This ensures that all stakeholders have a common understanding of the stability and usability of each version.

**Reply by sithlord48** [reply]
> Don't ever rename or replace artifacts of tagged versions.

Yeah we will replace them if we need to in the case of a packaging error however I do agree they should not be renamed anymore. the rename was to ensure we do not have to do it again with our tagged release this why uses can find old version more easily then if we had the old naming style on one single release.

**Reply by marlop352** [reply]
> > Don't ever rename or replace artifacts of tagged versions.
> 
> Yeah we will replace them if we need to in the case of a packaging error however I do agree they should not be renamed anymore. the rename was to ensure we do not have to do it again with our tagged release this why uses can find old version more easily then if we had the old naming style on one single release.

Tag a new release to fix whatever error happened, and if you think its necessary delete the affected artifact but **do not** replace a released artifact.

**Comment by sithlord48** [comment]
Just because a release is tagged does not make it stable you all need to stop working under this assumption.

**Reply by fuzzzerd** [reply]
The issue is not the git tag being stable or not, the issue is the GitHub release being tagged as latest and not pre-release. Everyone accepts that a given tag maybe unstable, but then if its associated with a release, that release should have the GitHub pre-release flag set true.

**Reply by sithlord48** [reply]
the problem is that github auto tags latest to the most recent released tag  this is again why no one should be expecting it to be the newest version at all do not use the `latest` tag its not always pointing where you think but `v.1.17.0` is

**Reply by marlop352** [reply]
Then fix the release process so that the continuous tag is always marked as pre-release, then latest will point to the correct place.

You seem to be the only one that wants to keep "continuous" not marked as pre-release regardless of how many people disagree with you.

Why don't you want to make this change at all? (consider that wanting users to use that release is not a valid reason)

**Reply by sithlord48** [reply]
because it is not a `pre-release` and making it so will move it to below the last tagged release on the release page. 

Update: I am incorrect here `pre-release` is not pushed down

**Comment by sithlord48** [comment]
Ideally I would like to prevent the release from tagging `continuous` as latest while also not making continuous a pre-release

**Reply by marlop352** [reply]
I'm gonna try and do some research on how but it should be possible considering this: https://github.com/cli/cli/issues/2678#issuecomment-1289018771

**Reply by marlop352** [reply]
this GH Action may do everything that you want: https://github.com/softprops/action-gh-release
if it doesn't it should be a good starting point for you to find one that does

**Comment by sithlord48** [comment]
I've made a PR https://github.com/deskflow/deskflow/pull/7839 that sums this up pretty well 
Reviews encouraged

**Reply by nbolton** [reply]
PR landed. Continuous is now "pre-release"

**Comment by nbolton** [comment]
Related but tangential discussion: #7843

**Reply by sithlord48** [reply]
please come see #7845 .  I think i have good solution for all here.
